src/json-patch-duplex.ts

```diff
@@ -524,95 +513,263 @@ module jsonpatch {
-      var obj = tree;
-      var t = 1; //skip empty element - http://jsperf.com/to-shift-or-not-to-shift
-      var len = keys.length;
-      var existingPathFragment = undefined;
+  export function unescapePathComponent(path: string): string {
+    return path.replace(/~1/g, '/').replace(/~0/g, '~');
+  }
 
+  /**
+   * Retrieves a value from a JSON document by a JSON pointer.
+   * Returns the value.
+   *
+   * @param document The document to get the value from
+   * @param pointer an escaped JSON pointer
+   * @return The retrieved value
+   */
+  export function getValueByPointer(document: any, pointer: string): any {
+    let pathSegments = [];
+    if (pointer) { //empty string evaluates to false
+      pathSegments = pointer.substring(1).split(/\//).map(jsonpatch.unescapePathComponent);
+    }
+    for (let i = 0; i < pathSegments.length; i++) {
+      let subPropertyKey = pathSegments[i];
+      if (!(subPropertyKey in document)) {
+        return undefined;
+      }
+      document = document[subPropertyKey];
+    }
+    return document;
+  }
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns the {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validateOperation `false` is without validation, `true` to use default jsonpatch's validation, or you can pass a `validateOperation` callback to be used for validation.
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
+   */
+  export function applyOperation<T>(document: T, operation: Operation, validateOperation: boolean | Validator<T> = false, mutateDocument: boolean = true): OperationResult<T> {
+    if (validateOperation) {
+      if(typeof validateOperation == 'function') {
+        validateOperation(operation, 0, document, operation.path);
+      }
+      else {      
+        validator(operation, 0);
+      }
+    }
+    const returnValue: OperationResult<T> = { newDocument: document, result: undefined };
+    /* ROOT OPERATIONS */
+    if (operation.path === "") {
+      if (operation.op === 'add') {
+        returnValue.newDocument = operation.value;
+        returnValue.result = undefined;
+        return returnValue;
+      } else if (operation.op === 'replace') {
+        returnValue.newDocument = operation.value;
+        returnValue.result = document; //document we removed
+        return returnValue;
+      }
+      else if (operation.op === 'move' || operation.op === 'copy') { // it's a move or copy to root
+        returnValue.newDocument = getValueByPointer(document, operation.from); // get the value by json-pointer in `from` field
+        if (operation.op === 'move') { // report removed item
+          returnValue.result = document;
+        }
+        return returnValue;
+      } else { // it's a remove or test on root
```

Comment:
is there a reason to use `else { if` instead of `else if` and reduce indentation for one level?